File: app/src/main/java/wangdaye/com/geometricweather/common/ui/widgets/DrawerLayout.java
package wangdaye.com.geometricweather.common.ui.widgets;

import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;
import android.widget.FrameLayout;

import androidx.annotation.FloatRange;
import androidx.annotation.Nullable;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.utils.DisplayUtils;

public class DrawerLayout extends ViewGroup {

    private @Nullable View mDrawer;
    private @Nullable View mContent;

    private boolean mUnfold;
    private @FloatRange(from = 0, to = 1) float mProgress; // 0 - fold, 1 - unfold.

    private @Nullable ValueAnimator mProgressAnimator;

    private static final int MIN_DRAWER_WIDTH_DP = 280;
    private static final int MAX_DRAWER_WIDTH_DP = 320;

    public DrawerLayout(Context context) {
        this(context, null);
    }

    public DrawerLayout(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public DrawerLayout(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.DrawerLayout, defStyleAttr, 0);
        mUnfold = a.getBoolean(R.styleable.DrawerLayout_unfold, false);
        mProgress = mUnfold ? 1 : 0;
        a.recycle();
    }

    @Override
    protected LayoutParams generateDefaultLayoutParams() {
        return new LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT);
    }

    @Override
    public LayoutParams generateLayoutParams(AttributeSet attrs) {
        return new LayoutParams(getContext(), attrs);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        if (getChildCount() > 0) {
            mDrawer = getChildAt(0);
        }
        if (getChildCount() > 1) {
            mContent = getChildAt(1);
        }

        LayoutParams lp;
        if (mDrawer != null) {
            lp = mDrawer.getLayoutParams();
            int width = lp.width;
            if (width == LayoutParams.WRAP_CONTENT) {
                width = getMeasuredWidth()
                        - DisplayUtils.getTabletListAdaptiveWidth(getContext(), getMeasuredWidth());
                if (width == 0) {
                    width = LayoutParams.MATCH_PARENT;
                } else {
                    int minDrawerWidth = (int) DisplayUtils.dpToPx(getContext(), MIN_DRAWER_WIDTH_DP);
                    int maxDrawerWidth = (int) DisplayUtils.dpToPx(getContext(), MAX_DRAWER_WIDTH_DP);
                    width = Math.max(width, minDrawerWidth);
                    width = Math.min(width, maxDrawerWidth);
                }
            }
            mDrawer.measure(
                    getChildMeasureSpec(widthMeasureSpec, 0, width),
                    getChildMeasureSpec(heightMeasureSpec, 0, lp.height)
            );

            if (mContent != null) {
                lp = mContent.getLayoutParams();
                if (mDrawer.getMeasuredWidth() == getMeasuredWidth()) {
                    mContent.measure(
                            getChildMeasureSpec(widthMeasureSpec, 0, lp.width),
                            getChildMeasureSpec(heightMeasureSpec, 0, lp.height)
                    );
                } else {
                    final int widthUsed = (int) (mDrawer.getMeasuredWidth() * mProgress);
                    mContent.measure(
                            getChildMeasureSpec(widthMeasureSpec, widthUsed, lp.width),
                            getChildMeasureSpec(heightMeasureSpec, 0, lp.height)
                    );
                }
            }
        }
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        if (getChildCount() > 0) {
            mDrawer = getChildAt(0);
        }
        if (getChildCount() > 1) {
            mContent = getChildAt(1);
        }

        if (DisplayUtils.isRtl(getContext())) {
            if (mDrawer != null) {
                mDrawer.layout(
                        (int) (getMeasuredWidth() - mDrawer.getMeasuredWidth() * mProgress),
                        0,
                        (int) (getMeasuredWidth() + mDrawer.getMeasuredWidth() * (1 - mProgress)),
                        mDrawer.getMeasuredHeight()
                );
            }
            if (mContent != null) {
                mContent.layout(
                        0,
                        0,
                        mDrawer.getLeft(),
                        mContent.getMeasuredHeight()
                );
            }
        } else {
            if (mDrawer != null) {
                mDrawer.layout(
                        (int) (mDrawer.getMeasuredWidth() * (mProgress - 1)),
                        0,
                        (int) (mDrawer.getMeasuredWidth() * mProgress),
                        mDrawer.getMeasuredHeight()
                );
            }
            if (mContent != null) {
                mContent.layout(
                        mDrawer.getRight(),
                        0,
                        mDrawer.getRight() + mContent.getMeasuredWidth(),
                        mContent.getMeasuredHeight()
                );
            }
        }
    }

    public boolean isUnfold() {
        return mUnfold;
    }

    public void setUnfold(boolean unfold) {
        if (mUnfold == unfold) {
            return;
        }

        mUnfold = unfold;

        if (mProgressAnimator != null) {
            mProgressAnimator.cancel();
            mProgressAnimator = null;
        }

        mProgressAnimator = generateProgressAnimator(mProgress, unfold ? 1 : 0);
        mProgressAnimator.start();
    }

    private ValueAnimator generateProgressAnimator(float from, float to) {
        ValueAnimator a = ValueAnimator.ofFloat(from, to);
        a.addUpdateListener(animation -> setProgress((Float) animation.getAnimatedValue()));
        a.setDuration((long) (Math.abs(from - to) * 450));
        a.setInterpolator(new DecelerateInterpolator(2f));
        return a;
    }

    private void setProgress(float progress) {
        mProgress = progress;
        requestLayout();
    }
}

File: app/src/main/java/org/breezyweather/ui/common/widgets/DrawerLayout.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.common.widgets

import android.animation.ValueAnimator
import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.view.animation.DecelerateInterpolator
import android.widget.FrameLayout
import androidx.annotation.FloatRange
import androidx.core.view.isNotEmpty
import org.breezyweather.R
import org.breezyweather.common.extensions.dpToPx
import org.breezyweather.common.extensions.getTabletListAdaptiveWidth
import org.breezyweather.common.extensions.isLandscape
import org.breezyweather.common.extensions.isRtl
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class DrawerLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0,
) : ViewGroup(context, attrs, defStyleAttr) {
    private var mDrawer: View? = null
    private var mContent: View? = null
    private var mUnfold: Boolean

    // 0.0 - fold, 1.0 - unfold
    @FloatRange(from = 0.0, to = 1.0)
    private var mProgress: Float
    private var mProgressAnimator: ValueAnimator? = null

    init {
        val a = context.obtainStyledAttributes(attrs, R.styleable.DrawerLayout, defStyleAttr, 0)
        mUnfold = a.getBoolean(R.styleable.DrawerLayout_unfold, false) && context.isLandscape
        mProgress = if (mUnfold) 1f else 0f
        a.recycle()
    }

    override fun generateDefaultLayoutParams(): LayoutParams {
        return LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT)
    }

    override fun generateLayoutParams(attrs: AttributeSet): LayoutParams {
        return LayoutParams(context, attrs)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        if (isNotEmpty()) {
            mDrawer = getChildAt(0)
        }
        if (childCount > 1) {
            mContent = getChildAt(1)
        }
        var lp: LayoutParams
        mDrawer?.let { drawer ->
            lp = drawer.layoutParams
            var width = lp.width
            if (width == LayoutParams.WRAP_CONTENT) {
                width = measuredWidth - context.getTabletListAdaptiveWidth(measuredWidth)
                if (width == 0) {
                    width = LayoutParams.MATCH_PARENT
                } else {
                    val minDrawerWidth = context.dpToPx(MIN_DRAWER_WIDTH_DP.toFloat()).toInt()
                    val maxDrawerWidth = context.dpToPx(MAX_DRAWER_WIDTH_DP.toFloat()).toInt()
                    width = max(width, minDrawerWidth)
                    width = min(width, maxDrawerWidth)
                }
            }
            drawer.measure(
                getChildMeasureSpec(widthMeasureSpec, 0, width),
                getChildMeasureSpec(heightMeasureSpec, 0, lp.height)
            )
            mContent?.let { content ->
                lp = content.layoutParams
                if (drawer.measuredWidth == measuredWidth) {
                    content.measure(
                        getChildMeasureSpec(widthMeasureSpec, 0, lp.width),
                        getChildMeasureSpec(heightMeasureSpec, 0, lp.height)
                    )
                } else {
                    val widthUsed = (drawer.measuredWidth * mProgress).toInt()
                    content.measure(
                        getChildMeasureSpec(widthMeasureSpec, widthUsed, lp.width),
                        getChildMeasureSpec(heightMeasureSpec, 0, lp.height)
                    )
                }
            }
        }
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        if (isNotEmpty()) {
            mDrawer = getChildAt(0)
        }
        if (childCount > 1) {
            mContent = getChildAt(1)
        }
        if (context.isRtl) {
            mDrawer?.let { drawer ->
                drawer.layout(
                    (measuredWidth - drawer.measuredWidth * mProgress).toInt(),
                    0,
                    (measuredWidth + drawer.measuredWidth * (1 - mProgress)).toInt(),
                    drawer.measuredHeight
                )
                mContent?.let { content ->
                    content.layout(
                        0,
                        0,
                        drawer.left,
                        content.measuredHeight
                    )
                }
            }
        } else {
            mDrawer?.let { drawer ->
                drawer.layout(
                    (drawer.measuredWidth * (mProgress - 1)).toInt(),
                    0,
                    (drawer.measuredWidth * mProgress).toInt(),
                    drawer.measuredHeight
                )
                mContent?.let { content ->
                    content.layout(
                        drawer.right,
                        0,
                        drawer.right + content.measuredWidth,
                        content.measuredHeight
                    )
                }
            }
        }
    }

    var isUnfold: Boolean
        get() = mUnfold
        set(unfold) {
            if (mUnfold == unfold) {
                return
            }
            mUnfold = unfold
            mProgressAnimator?.let {
                it.cancel()
                mProgressAnimator = null
            }
            mProgressAnimator = generateProgressAnimator(mProgress, if (unfold) 1f else 0f).also { it.start() }
        }

    private fun generateProgressAnimator(from: Float, to: Float): ValueAnimator {
        return ValueAnimator.ofFloat(from, to).apply {
            addUpdateListener { animation: ValueAnimator -> setProgress(animation.animatedValue as Float) }
            duration = (abs(from - to) * 450).toLong()
            interpolator = DecelerateInterpolator(2f)
        }
    }

    private fun setProgress(progress: Float) {
        mProgress = progress
        requestLayout()
    }

    companion object {
        private const val MIN_DRAWER_WIDTH_DP = 280
        private const val MAX_DRAWER_WIDTH_DP = 320
    }
}
